### ISSUE TITLE
Go indexer: More marked source cleanup. [phab:D1685]

### ISSUE THREAD

**Issue by kythe** [google/kythe#2110]
_Created by **fromberger** at 2017-05-31 00:13:47_: 

When rendering context, ensure that separators are included only when the
context is not empty. Otherwise a local name like "foo" may be rendered as
".foo" when it is shown without context (e.g., when a context is present but is
not being displayed by the rendering chosen).

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-05-31 02:32:44_:

This renders as
```
child {
  pre_text: "type "
}
child {
  child {
    kind: CONTEXT
    child {
      pre_text: "structtype"
    }
    post_child_text: "."
  }
  child {
    kind: IDENTIFIER
    pre_text: "T"
  }
  post_child_text: "."
}
child {
  pre_text: " "
}
child {
  kind: TYPE
  pre_text: "struct {...}"
}
```
which will draw as
```
      RenderSimpleIdentifier: "T"
RenderSimpleQualifiedName-ID: "structtype"
RenderSimpleQualifiedName+ID: "structtypeT"
```
according to the `doc` tool. I suggest adding add_final_list_token to the CONTEXT:
```
child {
  pre_text: "type "
}
child {
  child {
    kind: CONTEXT
    child {
      pre_text: "structtype"
    }
    post_child_text: "."
    add_final_list_token: true
  }
  child {
    kind: IDENTIFIER
    pre_text: "T"
  }
  post_child_text: "."
}
child {
  pre_text: " "
}
child {
  kind: TYPE
  pre_text: "struct {...}"
}
```
which will render as "structtype.T" where appropriate.

There might not yet be a way to get the full signature to render nicely in all contexts (where you can see `struct {...}`); I also don't believe there are any extant consumers that are interested in rendering any text dominated only by top-level BOXes. The best thing to do immediately is to add a space before "struct {...}" in the TYPE node and to eliminate the previous child. I think we need some sort of flexible space that's more general than add_final_list_token.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-05-31 02:47:37_:

That's what I originally had, but then the end-to-end test doubles up the ".". 

So which piece of the rendering machinery is wrong here? And do you have any suggestions how I can fix it?

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2017-05-31 02:59:56_:

I had previously labelled this whole subtree as TYPE, but that //really// confused the renderer. I backed out that change, but it sounds like maybe I went too far. I'm reluctant to keep tweaking randomly here; it might be better for the short term if I just dump everything except bare identifiers.

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-05-31 04:42:05_:

  - Include spaces in the type tags.

**Comment by kythe** [comment]
_**fromberger** wrote at 2017-05-31 20:16:58_:

  - Ensure contexts have a final list token.

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2017-05-31 20:33:43_

**Comment by kythe** [comment]
_**fromberger** closed via commit 087c2670503160fa9fac598660638177cfb32655 at 2017-05-31 21:57:13_
